### DISCUSSION TITLE
Possible bug in Core-metadata / EdgeX C SDK  file metadata.c function edgex_metadata_client_add_or_modify_device()

### DISCUSSION THREAD

**Discussion by satyam1990** [edgexfoundry/community#77] [General]
The function edgex_metadata_client_add_or_modify_device() does not works as expected. Reason stated below:

This function is supposed to add or update an existing device in core-metadata. But the function is unable to update device, reason being.

First in this function createdevice request is sent to core-metadata and it is assumed that if device already exists core-metadata will send HTTP 409, as written in metadata.c:
```
 json = edgex_createdevicereq_write (dev);
 if (edgex_http_post (lc, &ctx, url, json, edgex_http_write_cb, &err) == 409)
 {
     // Send Patch Request to update device
 }
```

But I could see core-metadata instead returns HTTP 207 Multi-Status. But if one looks at HTTP Response Text it is something as below:
```
[
    {
        "apiVersion": "v2",
        "message": "device name Temp-Device01 already exists",
        "statusCode": 409
    }
]
```
Due to this above if check never returns true and PATCH request is never initiated.

Hence, either we need to
- update the SDK Code to check the "statusCode" inside response text, if it is 409 then send patch request.
- or we need to check for HTTP code 207 instead of 409 since that is what core-metadata sends when device already exists
- or we need to update core-metadata to return 409 as HTTP code

EdgeX foundry version I am using is: 2.1.1
EdgeX C SDK I am using is: 2.2.0

For now I have modified C SDK to make things work but I guess it would be better if edgeX team can incorporate the fix if not already done.
Please do let me know if i need to raise a bug request for the same.

**Comment by iain-anderson** [comment]
Yes, it looks like the add-or-modify function hasn't been updated to fully take into account the move to a multiple-object call at that endpoint. Please file an issue against device-sdk-c

**Comment by satyam1990** [comment]
Thanks @iain-anderson . Raised bug request for the same.
https://github.com/edgexfoundry/device-sdk-c/issues/452
